devito/types/tensor.py

```diff
@@ -0,0 +1,601 @@
+        return self.__getitem__(i, j)
+
+    def __getitem__(self, *args):
+        if len(args) == 1:
+            return super(TensorFunction, self).__getitem__(*args)
+        i, j = args
+        if self.is_diagonal:
+            if i == j:
+                return super(TensorFunction, self).__getitem__(i, j)
+            return 0.0
+        if self.is_symmetric:
+            if j < i:
+                return super(TensorFunction, self).__getitem__(j, i)
+            else:
+                return super(TensorFunction, self).__getitem__(i, j)
+        return super(TensorFunction, self).__getitem__(i, j)
+
+    def _eval_transpose(self):
+        if self.is_symmetric or self.is_diagonal:
+            return self
+        mat = [[self[j, i] for j in range(self.cols)] for i in range(self.rows)]
+        func = tens_func(self, self)
+        name = '%s_T' % self.name
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=mat, time_order=to, symmetric=self.is_symmetric,
+                    diagonal=self.is_diagonal)
+
+    def values(self, symmetric=False):
+        if self.is_diagonal:
+            return [self[i, i] for i in range(self.shape[0])]
+        elif self.is_symmetric:
+            val = super(TensorFunction, self).values()
+            return list(OrderedDict.fromkeys(val))
+        else:
+            val = super(TensorFunction, self).values()
+            if symmetric:
+                val = [self[i, j] for j in range(self.cols)
+                       for i in range(self.rows) if j >= i]
+            return val
+
+    def _eval_add(self, other):
+        mat = [[self[i, j] + other[i, j] for j in range(self.cols)]
+               for i in range(self.rows)]
+        func = tens_func(self, other)
+        name = "%s%s" % (self.name, getattr(other, 'name', str(other)))
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=mat, time_order=to, symmetric=self.is_symmetric,
+                    diagonal=self.is_diagonal)
+
+    def _eval_scalar_mul(self, other):
+        mat = [[self[i, j]*other for j in range(self.cols)] for i in range(self.rows)]
+        func = tens_func(self, other)
+        name = "%s%s" % (self.name, getattr(other, 'name', str(other)))
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=mat, time_order=to, symmetric=self.is_symmetric,
+                    diagonal=self.is_diagonal)
+
+    def _eval_scalar_rmul(self, other):
+        return self._eval_scalar_mul(other)
+
+    @property
+    def div(self):
+        """
+        Divergence of the TensorFunction (is a VectorFunction).
+        """
+        comps = []
+        to = getattr(self, 'time_order', 0)
+        func = vec_func(self, self)
```

Comment:
This is a bit weird. Possible to clean up the `vec_func` method?

Reply:
what do you mean?

Reply:
`vec_func` is checking the same thing twice here.